thunder/torch/__init__.py

```diff
@@ -5331,6 +5331,230 @@ def wait(slf) -> None:
         utils.check(False, lambda: "torch.distributed is not available")
 
 
+#
+# The automatically registered torch operators
+#
+
+
+def _is_differentiable(arg: Any) -> bool:
+    from torch._subclasses.fake_tensor import FakeTensor
+
+    if isinstance(arg, (torch.Tensor, FakeTensor, TensorProxy)):
+        return dtypes.is_inexact_dtype(to_dtype(arg.dtype))
+    return False
+
+
+def _make_differentiable_wrapper(func: Callable, *args, **kwargs):
+    from thunder.core.pytree import tree_flatten, tree_unflatten
+
+    flat_args, spec = tree_flatten((args, kwargs))
+    differentiable_args_idx = tuple(i for i, a in enumerate(flat_args) if _is_differentiable(a))
+    differentiable_args = tuple(arg for i, arg in enumerate(flat_args) if i in differentiable_args_idx)
+
+    def wrapper(*diff_args):
+        diff_args_iter = iter(diff_args)
+        full_args = [next(diff_args_iter) if i in differentiable_args_idx else arg for i, arg in enumerate(flat_args)]
+        full_args, full_kwargs = tree_unflatten(full_args, spec)
+        return func(*full_args, **full_kwargs)
+
+    return wrapper, differentiable_args
+
+
+def register_default_torch_ops():
+    from thunder.torch import default_torch_ops
+
+    for m, fns in default_torch_ops.torch_auto_registered_ops.items():
+        for fn in fns:
+            register_default_torch_op(fn, meta_adaptor(fn), m)
+
+
+def register_default_torch_op(torchfn: Callable, fn_meta: Callable, m):
+    _fn = langctx(Languages.TORCH)(fn_meta)
+    sym = Symbol(
+        name=torchfn.__name__,
+        meta=_fn,
+        id=f"{m.__name__}.{torchfn.__name__}",
+    )
+    _torch_to_thunder_function_map[torchfn] = sym
+    from thunder.executors.torchex import _always_executable, ex
+
+    op = ex.register_operator(torchfn.__name__, module=m, meta=fn_meta)
+    ex.register_implementation(sym, op, checker=_always_executable)
+
+    from thunder.core.transforms import augmented_forward_impls, backward_impls
+
+    augmented_forward_impls[sym.id] = augmented_forward_adaptor(op)
+
+    def _vjp_impl(residules, *gs) -> torch.Tensor:
```

Comment:
It seems the impl func is not always near the metafunction, I thought adding another layer of wrapper just to move the impl near the metafunc might be unnecessary, what do you think?

Reply:
I don't have a strong preference to move them closer but here is why I think it would be nice to have them together.
The function bodies look similar and if we need an update then it is likely that both need to be updated - so having them close would probably remind this sooner. (But again not a strong preference, so feel free to resolve this if you think it would make the code harder to follow)

Reply:
Yeah, you convinced me :), "if we need an update then it is likely that both need to be updated", a simple wrapper func can't hurt, I'll update it